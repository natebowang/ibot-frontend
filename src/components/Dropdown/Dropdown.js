import React, {useState, useEffect, useCallback} from "react";

/**
 * Dropdown menu
 * @param {Component} toggleButtonIcon - An arrow or an list button
 * @param {style} toggleButtonStyle
 * @param children - the list to be open or closed
 * @returns {*}
 */
const Dropdown = ({toggleButtonIcon, toggleButtonStyle, children}) => {
    const [ifOpen, setIfOpen] = useState(false);

    const windowClickHandler = useCallback((e) => {
        e.preventDefault();
        setIfOpen(false);
        window.removeEventListener('click', windowClickHandler);
    }, []);

    useEffect(() => {
        return () => {
            window.removeEventListener('click', windowClickHandler);
        }
    }, []);

    const toggleHandler = (e) => {
        e.preventDefault();
        switch (ifOpen) {
            case true:
                setIfOpen(!ifOpen);
                window.removeEventListener('click', windowClickHandler);
                break;
            case false:
                setIfOpen(!ifOpen);
                setTimeout(() => {
                    window.addEventListener('click', windowClickHandler);
                }, 0);
                break;
            default:
                break;
        }
    };

    return (
        <div>
            <button
                onClick={toggleHandler}
                style={toggleButtonStyle}
            >
                {toggleButtonIcon}
            </button>
            {ifOpen ? children : null}
        </div>
    );
};

export default Dropdown;

export const DropdownExample = () => (
    <Dropdown
        toggleButtonIcon={<div>v</div>}
        toggleButtonStyle={toggleButtonStyle}
    >
        <ul style={ulStyle}>
            <li>1</li>
            <li>2</li>
            <li>3</li>
        </ul>
    </Dropdown>
);

const toggleButtonStyle = {
    backgroundColor: 'green',
};

const ulStyle = {
    background: 'yellow',
    margin: '0',
};
