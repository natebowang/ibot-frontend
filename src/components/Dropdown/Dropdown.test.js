import React from "react";
import renderer from 'react-test-renderer';
import {DropdownExample} from "./Dropdown";

test('Dropdown menu snapshot test', () => {
    const dropdown = renderer.create(
        <DropdownExample/>
    );

    const tree = dropdown.toJSON();
    expect(tree).toMatchSnapshot(tree);
});

