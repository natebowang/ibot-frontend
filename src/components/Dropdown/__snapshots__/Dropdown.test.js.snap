// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown menu snapshot test 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            "v",
          ],
          "props": Object {},
          "type": "div",
        },
      ],
      "props": Object {
        "onClick": [Function],
        "style": Object {
          "backgroundColor": "green",
        },
      },
      "type": "button",
    },
  ],
  "props": Object {},
  "type": "div",
}
`;
